src/fges.jl

```diff
@@ -0,0 +1,408 @@
+
+    verbose && println("Start forward search")
+    # Perform the forward search 
+    
+    ges_search!(g, scoreT, data, Insert!, verbose)
+
+    verbose && println("Start backward search")
+    # Perform the backward search 
+    ges_search!(g, scoreT, data, Delete!, verbose)
+    
+    return g # Return the graph
+
+end
+
+####################################################################
+# Insert and Delete Operators
+####################################################################
+"""
+
+    Insert!(g, x, y, T)
+
+Inserts x->y and directs previously undirected edges t->y, t ∈ T.
+Here x and y are not adjacent and T are undirected-neighbors of y 
+that are not adjacent to x.
+"""
+function Insert!(g, x, y, T)
+    add_edge!(g, x → y)
+    
+    # Orient all edges in T incident into child node
+    for t ∈ T
+        orientedge!(g, t → y)
+    end
+    return g
+end
+
+
+
+"""
+
+    Deletes!(g, x, y, H)
+
+Deletes x-y or x->y and directs previously undirected edges x->h and y->h
+for h in H.
+"""
+function Delete!(g, x, y, H)
+
+    # Remove directed or undirected edges (x→y and x-y)
+    rem_edge!(g, x → y)
+    rem_edge!(g, y → x)
+    
+    # Orient all vertices in H toward x and y
+    for h ∈ H
+        if has_both(g, x, h) # reading literally Definition 13, Chickering
+            orientedge!(g, x → h) 
+        end
+        orientedge!(g, y → h)
+    end
+
+    return nothing
+end
+
+function Insert!(g, nextstep::Step) 
+    x, y = Pair(nextstep.edge)
+    T = nextstep.subset
+    Insert!(g, x, y, T)
+end
+function Delete!(g, nextstep::Step) 
+    x, y = Pair(nextstep.edge)
+    H = nextstep.subset
+    Delete!(g, x, y, H)
+end
+
+####################################################################
+# Forward and Backward search
+####################################################################
+
+function ges_search!(g, scoreT, data, operator, verbose)
+
+    # Continually add/remove edges to the graph until the score stops increasing
+    while true
+        
+        # Get the new best step (depends on if we're inserting or deleting)
+        if operator == Insert!
+            step = find_insert(scoreT, data, g, verbose)
+        else
+            step = find_delete(scoreT, data, g, verbose)
+        end
+        
+        
+        # If the score did not improve...
+        if step.Δscore ≤ 0
+            #...stop searching
+            break
+        end
+        
+        # Use the insert or delete operator update the graph
+        operator(g, step)
+        
+        # Convert the PDAG to a complete PDAG
+        # Undirect all edges unless they participate in a v-structure
+        vskel!(g)
+        # Apply the 4 Meek rules to orient some edges in the graph
+        meek_rules!(g)
+    end
+
+    return nothing
+end
+
+
+function find_insert(scoreT, data, g, verbose)
+    nextstep = Step{Int,scoreT}()
+    # Loop through all possible node combinations
+    for (x,y) in product(vertices(g), vertices(g)) 
```

Comment:
... and this one:  Both bugs needed to be fixed simultaneously for the test not to break..